### DISCUSSION TITLE
Using parser with base64 encoded query string

### DISCUSSION THREAD

**Discussion by kderck** [aws-powertools/powertools-lambda-python#2575] [Q&A]
Hi, 

I have a string that is a `base64 encoded query string` is there anyway I can use the parser function to nicely parse this into class without excessive boilerplate?

```python
class Message:
  msg: string 
  user_id: string 
  ...
````

**Comment by leandrodamascena** [comment]
Hi @kderck! Could you please elaborate more on this? Are you trying to use our Parser utility for this?

Thank you.

**Reply by kderck** [reply]
Hi @leandrodamascena. Unfortunately. I have legacy payload from Slack. The event comes in as JSON, However, this event has encoded the value of the body as base64. I'd like to be able to parse the body in the event 

```json
{
  "headers": {"client_ip":  "192.169.0.1"}
 "body": "base64 encoded value"
}
```

**Reply by heitorlessa** [reply]
What @leandrodamascena was after was a reproducible info or snippet to offer a more prescriptive guidance :)  e.g., how your function gets invoked, sample payload, whether you're using Powertools Parser (Pydantic) etc.

**TL;DR**: You can use JMESPath Utilities to use our `powertools_base64` along with `powertools_json` to decode and deserialize query strings back.

**Assuming** you meant the encoded query string once decoded looks like a dictionary:

> b64 JSON value: `eyJtc2ciOiAiaGVsbG8iLCAidXNlcl9pZCI6ICJkY2M2ZDAzMS00MmJjLTQ0MmUtYmIwYS1lMWJhNzk2YmNkMWMifQ==`

```python
{
    "msg": "hello",
    "user_id": "dcc6d031-42bc-442e-bb0a-e1ba796bcd1c"
}
```

You can reduce the common boilerplate of `decode/encode/json.loads) by using JMESPath Utility. There are built-in functions like `powertools_base64` and `powertools_json` that when combined leads to the result you want.

**Reproducible example**

```python
import base64
import json
from aws_lambda_powertools.utilities.jmespath_utils import extract_data_from_envelope
from aws_lambda_powertools.utilities.parser import BaseModel

# Reusing your Message class as a Pydantic model... Python @dataclass also works
class Message(BaseModel):
    msg: str
    user_id: str

# Raw dict query string as a JSON string
raw_qs = json.dumps({"msg": "hello", "user_id": "dcc6d031-42bc-442e-bb0a-e1ba796bcd1c"})

# Base64 encoded JSON query string as string
b64_encoded_qs: str = base64.b64encode(raw_qs.encode()).decode()

# Mimicking the payload you shared
event = {
    "body": b64_encoded_qs
}

## What You Actually Need
#
# 1. Decode base64 string available in `body` key
# 2. Deserialize JSON string into a normal dict
decoded_qs = extract_data_from_envelope(data=event, envelope="powertools_json( powertools_base64(body) )")

# Use expansion to convert dict into key=value arguments to instantiate Message class
message = Message(**decoded_qs)

assert message.msg == "hello"
```

PS: The name `extract_data_from_envelope` is an old tech debt we haven't addressed yet, we should have a more friendly name.

**Reply by heitorlessa** [reply]
If you are however using API Gateway as an event source, you can use [Event Handler feature](https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/) to further reduce boilerplate for handling a specific webhook request... then use `app.current_event.decoded_body` to get a string representation.

> That's why @leandrodamascena asked for more information, as there's a quick way to solve the problem (JMESPath) and there's the elegant way with Event Handler ;)

```python
from aws_lambda_powertools import Logger
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.logging import correlation_paths
from aws_lambda_powertools.utilities.typing import LambdaContext

logger = Logger()
app = APIGatewayRestResolver()


class Message(BaseModel):
    msg: str
    user_id: str


@app.get("/slack_webhook")
@tracer.capture_method
def get_todos():
    payload = json.loads(app.current_event.decoded_body)
    message = Message(**payload)

    return {"message": message.msg}


# You can continue to use other utilities just as before
@logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST)
def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

**Reply by kderck** [reply]
Hi @heitorlessa thank for your incredibly well wrote answer. I appreciate it, and I apologise for the lack of in-depth context in the request. I'l follow up after I've given it ago!

**Reply by heitorlessa** [reply]
No worries at all ;) Looking forward to hearing from you

**Reply by kderck** [reply]
Hi @heitorlessa @leandrodamascena haven't forgotten just haven't got back to this particular lambda yet.

**Comment by leandrodamascena** [comment]
Closing this discussion as it is resolved.

Feel free to reopen if you have more questions @kderck.
